### PR TITLE
Fixed paging in OpenId by using a dynamic shapefactory

### PR DESCRIPTION
**Pull request by MatthijsKrempel**
Fix for issue https://github.com/OrchardCMS/OrchardCore/issues/4396

### CODE REVIEW

**Comment by PinpointTownes** [comment]
Thanks for the PR! As I'm not super familiar with the shaper thing, I'll let @sebastienros take a look 😄

File: src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs [hunk 1]
```diff
@@ -65,13 +66,11 @@ public async Task<ActionResult> Index(PagerParameters pagerParameters)
 
             var siteSettings = await _siteService.GetSiteSettingsAsync();
             var pager = new Pager(pagerParameters, siteSettings.PageSize);
+            var count = await _applicationManager.CountAsync();
 
-            var model = new OpenIdApplicationsIndexViewModel
+            var model = new OpenIdApplicationsIndexViewModel()
```
[the remark this answers is not in the document]
Comment (MatthijsKrempel) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs]: If I add one at the end, I get a compilation error, so I am not sure what you mean,.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ScopeController.cs [hunk 2]
```diff
@@ -58,6 +57,8 @@ public class ScopeController : Controller
             _shellHost = shellHost;
         }
 
+        public dynamic New { get; set; }
```
[the remark this answers is not in the document]
Comment (MatthijsKrempel) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ScopeController.cs]: I took this approach from the Admin section, I can make it a private or a getter only.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs]: No no, it is my bad, even the parathesis are not required here, I revoke some of my suggestions

[the remark this answers is not in the document]
Comment (MatthijsKrempel) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs]: I can revoke the parenthesis if you want?

[the remark this answers is not in the document]
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs]: You can do it, coz there 're not needed here

**Comment by hishamco** [comment]
Overall it LGTM, @jptissot do you have any comment on this one

File: src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs [hunk 3]
```diff
@@ -56,6 +55,8 @@ public class ApplicationController : Controller
             _shellDescriptor = shellDescriptor;
         }
 
+        public dynamic New { get; }
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs]: ShapeFactory? Or is it consistent with what we doing by using New?

Comment (jptissot) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs]: I've seen this pattern very often in the code.

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs]: Ya

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs]: Maybe inspired of our `RazorPage.New` property often used in razor views templates

        public dynamic New
        {
            get
            {
                return Factory;
            }
        }

Comment (MatthijsKrempel) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs]: Do we want to use a different approach?
Or can we accept this PR? :)

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs]: IMHO what's done now is fine, so :shipit:

Comment (MatthijsKrempel) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs]: Then I will resolve

Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/ApplicationController.cs]: It's fine ;)

**Comment by sebastienros** [comment]
I would suggest to prefer the PAgerSlim as it won't require a Count call. In this case I think it's fine but if there are many elements then it could be a big perf issue to count them (table scan).

**Pull request merged**